### PR TITLE
8312383: Log X509ExtendedKeyManager implementation class name in TLS/SSL connection

### PR DESCRIPTION
**Pull request by pkumaraswamy**
During the time of server certificate validation, users have the flexibility to use a custom X509 Key Manager implementation by extending "X509ExtendedKeyManager.".
In such cases, printing the class name in X509Authentication.java will be helpful to trace any failure of the SSL connection due to a certificate issue.

I've tested the code by running the custom X509 manager, the default X509 manager, and passing the null key manager.
The screen shots are attached here.
[x509_log_testing.zip](https://github.com/openjdk/jdk/files/14206695/x509_log_testing.zip)

Also, the internal test runs against this fix are green

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312383](https://bugs.openjdk.org/browse/JDK-8312383): Log X509ExtendedKeyManager implementation class name in TLS/SSL connection (**Enhancement** - P4)

### Reviewers
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17742/head:pull/17742` \
`$ git checkout pull/17742`

Update a local copy of the PR: \
`$ git checkout pull/17742` \
`$ git pull https://git.openjdk.org/jdk.git pull/17742/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17742`

View PR using the GUI difftool: \
`$ git pr show -t 17742`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17742.diff">https://git.openjdk.org/jdk/pull/17742.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17742#issuecomment-1931386405)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/ssl/X509Authentication.java [hunk 1]
```diff
@@ -201,6 +201,10 @@ public static SSLPossession createPossession(
     private static SSLPossession createClientPossession(
             ClientHandshakeContext chc, String[] keyTypes) {
         X509ExtendedKeyManager km = chc.sslContext.getX509KeyManager();
+        if (SSLLogger.isOn && SSLLogger.isOn("ssl")) {
+            SSLLogger.finest("X509ExtendedKeyManager being used: " +
```
Comment (coffeys) [inline src/java.base/share/classes/sun/security/ssl/X509Authentication.java:205]: Could the JBS title be made more descriptive ? It's quite vague.

I wonder if "X509KeyManager class: " would be better for displaying.

`createServerPossession` would also benefit from this logging enhancement. I wonder if this belongs in logging during SSLContext creation time instead. Other security-dev engineers may have opinion on that.

IIRC, there's another issue open where we iterate over the certificate contexts of custom tm/km types. The JDK src does it at the moment for the default tm/km but no output given for custom impl. Will be good to have that tied up at some stage also.

File: src/java.base/share/classes/sun/security/ssl/X509Authentication.java [hunk 2]
```diff
@@ -201,6 +201,10 @@ public static SSLPossession createPossession(
...
+                    (km == null ? "null" : km.getClass().getName()));
```
Comment (coffeys) [inline src/java.base/share/classes/sun/security/ssl/X509Authentication.java:206]: do you need to cater for null ? I thought a Dummy manager is returned in such scenarios.

Comment (pkumaraswamy) [reply src/java.base/share/classes/sun/security/ssl/X509Authentication.java:206]: No, I did verify that it always returns Dummy Manager even when key manager is initialized as null. I have added this as a safety net. I'll remove this.

File: src/java.base/share/classes/sun/security/ssl/X509Authentication.java (hunk 1 above)
Comment (pkumaraswamy) [reply src/java.base/share/classes/sun/security/ssl/X509Authentication.java:205]: I have made changes and attached the testing snapshots in a zip file attached in bug description

**Comment by pkumaraswamy** [comment]
Adding comment to avoid auto closure of PR

File: src/java.base/share/classes/sun/security/ssl/X509Authentication.java (hunk 1 above)
Comment (pkumaraswamy) [reply src/java.base/share/classes/sun/security/ssl/X509Authentication.java]: I have made changes to log statement

**Comment by pkumaraswamy** [comment]
I have tested the latest changes and attached is the snapshot
[Uploading x509kmprintlogs_v3.zip…]()

**Review by coffeys — APPROVED** [review]
Looks good.

**Pull request closed without merging**